### PR TITLE
Not allowing VstsCredentialProvider to be used if the URL is on prem

### PR DESCRIPTION
**Pull request by satbai**

### CODE REVIEW

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/BearerTokenProviders.cs [hunk 1]
```diff
@@ -1,153 +1,155 @@
...
+    /// <summary>
+    /// Acquire an AAD token via the ADAL cache
+    /// </summary>
+    public class AdalCacheBearerTokenProvider : IBearerTokenProvider
+    {
+        private readonly IAdalTokenProvider adalTokenProvider;
+
+        public AdalCacheBearerTokenProvider(IAdalTokenProvider adalTokenProvider)
+        {
             this.adalTokenProvider = adalTokenProvider;
```
Comment (satbai) [inline CredentialProvider.Microsoft/CredentialProviders/Vsts/BearerTokenProviders.cs:24]: VS complained about line endings :(

Comment (satbai) [reply CredentialProvider.Microsoft/CredentialProviders/Vsts/BearerTokenProviders.cs:24]: The real changes in this file are:
- added isOnPrem to ShouldRun
- WindowsIntegratedAuthBearerTokenProvider  doesn't run on prem

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs [hunk 2]
```diff
@@ -73,17 +81,25 @@ public async Task<Uri> GetAadAuthorityUriAsync(Uri uri, CancellationToken cancel
...
+            if (IsHttpsScheme(uri) && responseHeaders.Contains(VssResourceTenant) && responseHeaders.Contains(VssAuthorizationEndpoint))
             {
-                // We are not talking to a https endpoint so it cannot be a VSTS endpoint
-                return false;
+                return IFeedUriSource.Hosted;
             }
 
-            var responseHeaders = await GetResponseHeadersAsync(uri, cancellationToken: default);
+            // Uri returns 401 (TFS service error) if it's an Azure DevOps Server instance.
+            if (responseHeaders.Contains(TfsServiceError))
```
Comment (satbai) [inline CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs:96]: This is the header I'd like to have a discussion about

**Comment by satbai** [comment]
Logs;

[Verbose] [CredentialProvider]GET http://onpreminstance/DefaultCollection/one/_packaging/feed1/nuget/v3/index.json
[Verbose] [CredentialProvider]VstsCredentialProvider - Uri returned valid Azure DevOps headers.
[Verbose] [CredentialProvider]Using NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider to try get credentials for http://onpreminstance/DefaultCollection/one/_packaging/feed1/nuget/v3/index.json.
[Verbose] [CredentialProvider]IsRetry: True
[Verbose] [CredentialProvider]Invalidating SessionToken cache for http://onpreminstance/DefaultCollection/one/_packaging/feed1/nuget/v3/index.json
[Verbose] [CredentialProvider]GET http://onpreminstance/DefaultCollection/one/_packaging/feed1/nuget/v3/index.json
[Verbose] [CredentialProvider]Could not find AAD authority from headers, using default: https://login.microsoftonline.com
[Verbose] [CredentialProvider]VstsCredentialProvider - Using AAD authority: https://login.microsoftonline.com/common
[Verbose] [CredentialProvider]VstsCredentialProvider - Not running bearer token provider 'ADAL Cache'
[Verbose] [CredentialProvider]VstsCredentialProvider - Not running bearer token provider 'ADAL Windows Integrated Authentication'
[Verbose] [CredentialProvider]VstsCredentialProvider - Not running bearer token provider 'ADAL UI'
[Verbose] [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Device Code'
[Minimal] [CredentialProvider]DeviceFlow: http://onpreminstance/DefaultCollection/one/_packaging/feed1/nuget/v3/index.json
[Minimal] [CredentialProvider]ATTENTION: User interaction required.

    **********************************************************************

    To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code BSS7EUECU to authenticate.

    **********************************************************************

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs [hunk 3]
```diff
@@ -17,16 +17,24 @@ public interface IAuthUtil
     {
         Task<Uri> GetAadAuthorityUriAsync(Uri uri, CancellationToken cancellationToken);
 
-        Task<bool> IsVstsUriAsync(Uri uri);
+        Task<IFeedUriSource> GetFeedUriSource(Uri uri);
 
         Task<Uri> GetAuthorizationEndpoint(Uri uri, CancellationToken cancellationToken);
     }
 
+    public enum IFeedUriSource
```
Comment (jmyersmsft) [inline CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs:25]: * enum name shouldn't start with `I`
* Perhaps `AzDevDeploymentType` would be a clearer name?

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs (hunk 2 above)
Comment (jmyersmsft) [reply CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs:96]: I would use a different header. I don't think we want to specifically identify an error case here. `X-VSS-E2EID` should be on all responses

File: CredentialProvider.Microsoft/Resources.resx [hunk 4]
```diff
@@ -459,4 +459,17 @@ NuGet workarounds
   <data name="UIFlowStarted" xml:space="preserve">
     <value>Using the {0}  flow for uri {1}. User sign-in required in a pop-up authentication window.</value>
   </data>
+  <data name="HostAccepted" xml:space="preserve">
+    <value>Host accepted: {0}</value>
+  </data>
+  <data name="OnPremDetected" xml:space="preserve">
+    <value>Detected an on premise Azure DevOps Server. Please use a PAT and VSS_NUGET_EXTERNAL_FEED_ENDPOINTS variable to use this credential provider.</value>
```
Comment (jmyersmsft) [inline CredentialProvider.Microsoft/Resources.resx:466]:
```suggestion
    <value>Detected on-premises Azure DevOps Server.</value>
```
I don't think we need to tell customers about weird environment variables here. Customers should use regular nuget credential config mechanisms to set up their PATs

File: CredentialProvider.Microsoft/Resources.resx [hunk 5]
```diff
@@ -459,4 +459,17 @@ NuGet workarounds
   <data name="UIFlowStarted" xml:space="preserve">
     <value>Using the {0}  flow for uri {1}. User sign-in required in a pop-up authentication window.</value>
   </data>
+  <data name="HostAccepted" xml:space="preserve">
+    <value>Host accepted: {0}</value>
```
Comment (jmyersmsft) [inline CredentialProvider.Microsoft/Resources.resx:463]:
```suggestion
    <value>Matched well-known Azure DevOps Service hostname: {0}</value>
```

File: CredentialProvider.Microsoft/Resources.resx [hunk 6]
```diff
@@ -459,4 +459,17 @@ NuGet workarounds
...
+  </data>
+  <data name="OnPremDetected" xml:space="preserve">
+    <value>Detected an on premise Azure DevOps Server. Please use a PAT and VSS_NUGET_EXTERNAL_FEED_ENDPOINTS variable to use this credential provider.</value>
+  </data>
+  <data name="UsingCredentialProvider" xml:space="preserve">
+    <value>Using {0} to try get credentials for {1}.</value>
+    <comment>{0} - credential provider name, {1} - Uri</comment>
+  </data>
+  <data name="ValidHeaders" xml:space="preserve">
+    <value>Uri returned valid Azure DevOps headers.</value>
```
Comment (jmyersmsft) [inline CredentialProvider.Microsoft/Resources.resx:473]:
```suggestion
    <value>Detected hosted Azure DevOps Service.</value>
```

File: CredentialProvider.Microsoft/Resources.resx [hunk 7]
```diff
@@ -459,4 +459,17 @@ NuGet workarounds
...
     <value>Using the {0}  flow for uri {1}. User sign-in required in a pop-up authentication window.</value>
   </data>
+  <data name="HostAccepted" xml:space="preserve">
+    <value>Host accepted: {0}</value>
+  </data>
+  <data name="OnPremDetected" xml:space="preserve">
+    <value>Detected an on premise Azure DevOps Server. Please use a PAT and VSS_NUGET_EXTERNAL_FEED_ENDPOINTS variable to use this credential provider.</value>
+  </data>
+  <data name="UsingCredentialProvider" xml:space="preserve">
+    <value>Using {0} to try get credentials for {1}.</value>
```
Comment (jmyersmsft) [inline CredentialProvider.Microsoft/Resources.resx:469]:
```suggestion
    <value>Using {0} to try to get credentials for {1}.</value>
```

File: CredentialProvider.Microsoft/Resources.resx (hunk 4 above)
Comment (jmyersmsft) [reply CredentialProvider.Microsoft/Resources.resx:466]: With Aurelie's Connect-to-feed changes, I don't think we need to add any specific instructions here at all.

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/VstsCredentialProvider.cs [hunk 8]
```diff
@@ -49,16 +49,36 @@ public override async Task<bool> CanProvideCredentialsAsync(Uri uri)
...
+                Verbose(Resources.ValidHeaders);
+                return true;
+            }
+
+            if (feedUriSource == IFeedUriSource.OnPrem)
+            {
+                Verbose(Resources.OnPremDetected);
+                return false;
+            }
+
```
Comment (jmyersmsft) [inline CredentialProvider.Microsoft/CredentialProviders/Vsts/VstsCredentialProvider.cs:80]: Might want to add logging here for the `External` case, e.g. "Feed is not Azure DevOps."

**Pull request merged**